cache-filter/src/filter/http.rs

```diff
@@ -0,0 +1,125 @@
+        match get_application_from_cache(self.cache_key.as_str()) {
+            Some((mut app,_)) =>
+            {
+                info!("ctxt {}: Cache Hit", context_id);
+                if self.is_rate_limited(&request_data, &mut app, &current_time)
+                {
+                    info!("ctxt {}: Request is rate-limited", context_id);
+                    // Add some identifier for rate-limit filter
+                } 
+                else if self.to_report_singleton 
+                {
+                    info!("ctxt {}: Request is allowed to pass", context_id);
+                    if self.report_to_singleton(queue_id,&request_data) {
+                        // Handle MQ failure here
+                    }
+                }
+                return Action::Continue;
+            },
+
+            None =>
+            {
+                info!("ctxt {}: Cache Miss", context_id);
+                handle_cache_miss(&request_data);
+                return Action::Pause;
+            }
+        }
+    }
+
+}
+
+impl Context for CacheFilter {}
+
+impl CacheFilter {
+    fn report_to_singleton(&self, queue_id: Option<u32>, request_data: &ThreescaleData) -> bool {
+        match queue_id {
+            Some(qid) =>
+            {
+                if let Err(_) =  self.enqueue_shared_queue(qid, Some(&bincode::serialize(request_data).unwrap())) {
+                    info!("ctxt {}: Reporting to singleton failed: MQ with specified id not found", self.context_id);
+                    return true;
+                }
+            },
+            None =>
+            {
+                info!("ctxt {}: Reporting to singleton failed: Queue id not provided", self.context_id);
+                return true;
+            }
+        }
+        false
+    }
+
+    fn is_rate_limited(&mut self, request_data: &ThreescaleData, app: &mut Application, _current_time: &SystemTime) -> bool {
+        for (metric, hits) in request_data.metrics.borrow().iter() {
+            // Check metric is present inside local cache
+            if !app.local_state.borrow().contains_key(metric) {
+                continue;
+            }
+            
+            /* Check period window expiration
+            if app.local_state.borrow().get(metric).unwrap().period_window.end < current_time {
+                // Period window is expired
+                // Update period window using get_new_window for each metric
+                // But first confirm how to deal with time sync between host and 3scale
+                // Reset left_hits to max value
+            }*/
+    
+            // If any metric is rate-limited then whole request is restricted
+            if app.local_state.borrow().get(metric).unwrap().left_hits - hits < 0 {
+                true;
+            }
+        }
+
+        if !set_application_to_cache(self.cache_key.as_str(), app, true, None) {
+            self.to_report_singleton = true;
```

Comment:
This should be `false` since `set_application_to_cache` returns `false` when application is written to cache. Now the value of `to_report_singleton` is always `true` because it's initialized as `true` and here it's updated as `true`. I think it's better to change the `set_application_to_cache` logic so that it will return `true` when set application succeeds. Now with the inverted logic it's confusing and I think that's why you have mistakenly set up true here.

Reply:
Good catch! I'll change the return pattern and set the default value of `to_report_singleton` to `false`.